<a href="https://colab.research.google.com/github/vermelho25/analitica/blob/main/Reto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
#Para implementación y evaluación del modelo
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

In [ ]:
nxl = '/content/sample_data/2. TechSalesReport_Int_L.xlsx'
df = pd.read_excel(nxl,sheet_name=0)



In [ ]:
df2 = df[['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS','College']]

In [ ]:
df2['Business'] = df2['Business'].map({'Hardware': 0, 'Software': 1})
df2['College'] = df2['College'].map({'Yes': 1, 'No': 0})

<ipython-input-7-2fda56836961>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Business'] = df2['Business'].map({'Hardware': 0, 'Software': 1})
<ipython-input-7-2fda56836961>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['College'] = df2['College'].map({'Yes': 1, 'No': 0})


In [ ]:
XD=df2[['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS']]
yd=df2[['College']]

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
#Max Depth es el número de variables disponibles
#Gini es el criterio que indica el grado de error en el modelo

mar.fit(XD,yd)

#Se procede con el despliegue del arbol de decisión
from sklearn.tree import export_graphviz #Exportar
from pydotplus import graph_from_dot_data #Crear el gráfico

vs=['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS']
dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('1.TechSalesReport.png')

True

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

[[  114  4356]
 [   69 17451]]


In [ ]:
Exactitud=(VP+VN)/(VP+VN+FP+FN) #Funcionamiento general
print("La exactitud es:\n",Exactitud)
Sensibilidad=VP/(VP+FN) #Como se comporto el modelo con los aprobados
print("La sensibilidad es:\n",Sensibilidad)
Especificidad=VN/(VN+FP) #Como se comporto el modelo con los negados
print("La especificidad es:\n",Especificidad)
Precision=VP/(VP+FP) #Cual es el acierto clasificando los positivos
print("La precisión es:\n",Precision)
TasaError=(FP+FN)/(VP+VN+FP+FN) #Equivocación general
print("La tasa de error es:\n",TasaError)
PredNeg=VN/(VN+FN) #Como funciona el modelo negando créditos
print("La tasa de predicción negativa es:\n",PredNeg)

La exactitud es:
 0.7987721691678036
La sensibilidad es:
 0.9960616438356165
La especificidad es:
 0.025503355704697986
La precisión es:
 0.8002476269087908
La tasa de error es:
 0.20122783083219645
La tasa de predicción negativa es:
 0.6229508196721312


In [ ]:
# Se toma la totalidad de los solicitantes desde la nueva hoja con 16 datos
XDB2 = pd.read_excel(nxl, sheet_name=1)
XDB2.head(16)  # Solo se muestran los 16 datos

# Variable objetivo (College)
yd2 = XDB2[['College']]
yd2.head(16)

# Identificar las filas sin información en College
filas = np.where(np.isnan(yd2))[0]
print("Las filas que no poseen información:", filas)

# Seleccionar las variables para el pronóstico (sin incluir College)
XD2 = XDB2[['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS']]
XD2['Business'] = XDB2['Business'].map({'Hardware': 0, 'Software': 1})
XD2 = XD2.iloc[filas, :]  # Filtrar solo las filas sin datos en College


# Realizar el pronóstico con el árbol de decisión ya montado
ydp2=mar.predict(XD2)
print('Pronóstico:', ydp2)

# Crear el informe con los resultados
dfp = pd.DataFrame(np.column_stack((filas, XD2, ydp2)))
dfp.columns = ['Filas','Business','Age','Female','Years','Certificates','Feedback','Salary','NPS','College']
dfp.head(16)

# Guardar el informe en Excel
dfp.to_excel('InformeCollege.xlsx', index=False)
print("Las predicciones son:\n", ydp2)

# Contar los pronósticos de personas con y sin título universitario
print("Los sin título pronosticados son:", len(np.where(ydp2 == 0)[0]))
print("Los con título pronosticados son:", len(np.where(ydp2 == 1)[0]))

Las filas que no poseen información: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]
Pronóstico: [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
Las predicciones son:
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
Los sin título pronosticados son: 0
Los con título pronosticados son: 16


<ipython-input-22-c0e97c17113c>:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  XD2['Business'] = XDB2['Business'].map({'Hardware': 0, 'Software': 1})


Luego de la ejecución del modelo, se obtuvo una exactitud del 79.88%, lo que indica un desempeño aceptable en el pronóstico.

Es importante destacar que el coeficiente de sensibilidad fue 99.61%, lo que significa que el modelo tiene una alta capacidad para identificar correctamente a las personas que sí poseen un título universitario. Sin embargo, la especificidad fue 2.55%, lo que indica que el modelo tiene dificultades para identificar correctamente a las personas que no poseen un título.

Por otro lado, la precisión del modelo fue de 80.02%, lo que indica que, de todas las personas que fueron clasificadas como poseedoras de un título universitario, el 80% realmente lo tiene.

No obstante, la tasa de predicción negativa se ubicó en 62.29%, lo que sugiere que el modelo tiene una capacidad limitada para predecir correctamente a quienes no tienen un título universitario.

De la base de datos que había por pronosticar, los 16 individuos resultaron ser poseedores de títulos universitarios, absolutamente todos según el modelo.







